src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php

```diff
@@ -347,7 +367,7 @@ private function analyzeCircularReferences(array $edges, &$checkedNodes, &$curre
             $node = $edge->getDestNode();
             $id = $node->getId();
 
-            if ($node->getValue() && ($edge->isLazy() || $edge->isWeak())) {
+            if ($node->getValue() && (($edge->isLazy() && !$this->getProxyDumper() instanceof NullDumper) || $edge->isWeak())) {
```

Comment:
When no proxy dumper is set, we should lazy definitions should be involved into circular refs tracking.